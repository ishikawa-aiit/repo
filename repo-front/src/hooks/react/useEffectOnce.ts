import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
