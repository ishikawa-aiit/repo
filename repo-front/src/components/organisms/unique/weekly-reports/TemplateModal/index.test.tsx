import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { TemplateModal } from '.'

describe('props', () => {
  it('「反映する」ボタンをクリックすると、onApplyとonCloseが実行される', async () => {
    const onApply = jest.fn<() => void, never[]>()
    const onClose = jest.fn<() => void, never[]>()

    // eslint-disable-next-line @typescript-eslint/no-empty-function
    render(<TemplateModal isOpen={true} onClose={onClose} onApply={onApply} />)

    fireEvent.click(screen.getByTestId('TemplateModalApplyButton'))
    await waitFor(() => {
      expect(onClose).toBeCalled()
    })
    expect(onApply).toBeCalled()
  })

  it('「やめる」ボタンをクリックすると、onCloseが実行され、onApplyが実行されない', async () => {
    const onApply = jest.fn<() => void, never[]>()
    const onClose = jest.fn<() => void, never[]>()

    // eslint-disable-next-line @typescript-eslint/no-empty-function
    render(<TemplateModal isOpen={true} onClose={onClose} onApply={onApply} />)

    fireEvent.click(screen.getByTestId('TemplateModalCancelButton'))
    await waitFor(() => {
      expect(onClose).toBeCalled()
    })
    expect(onApply).not.toBeCalled()
  })
})

describe('snapshot', () => {
  it('最新のスナップショットに更新されている', () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    const { container } = render(<TemplateModal isOpen={true} onClose={() => {}} onApply={() => {}} />)

    expect(container).toMatchSnapshot()
  })
})
