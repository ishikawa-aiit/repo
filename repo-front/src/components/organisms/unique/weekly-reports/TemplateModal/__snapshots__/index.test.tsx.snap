// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 最新のスナップショットに更新されている 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
